pallets/inactivity-tracking/src/lib.rs

```diff
@@ -445,49 +447,80 @@ pub mod pallet {
+                let mut temp_set: BTreeSet<Collator<T>> = active_collators.clone().into();
+
+                for author in authors {
+                    temp_set.insert(author);
+                }
+
+                match BoundedBTreeSet::<Collator<T>, T::MaxCollatorsPerSession>::try_from(temp_set)
+                {
+                    Ok(bounded_set) => {
+                        *active_collators = bounded_set;
+                        Ok(())
                     }
-                    Ok(())
-                },
-            );
-            total_weight
+                    Err(_) => Err(()),
+                }
+            });
+
+            match result {
+                Ok(_) => {
+                    total_weight.saturating_accrue(T::DbWeight::get().writes(1));
+                    total_weight
+                }
+                Err(_) => {
+                    // If we reach MaxCollatorsPerSession limit there must be a bug in the pallet
+                    // so we disable the activity tracking
+                    Self::set_inactivity_tracking_status_inner(
+                        T::CurrentSessionIndex::session_index(),
+                        false,
+                    );
+                    total_weight.saturating_accrue(T::DbWeight::get().writes(2));
+                    total_weight
+                }
+            }
         }
 
         /// Internal update the current session active chains records.
         /// This function is called when a container chain is noted.
-        pub fn on_chain_noted(chain_id: ParaId) -> Weight {
-            let mut total_weight = T::DbWeight::get().reads_writes(1, 0);
-            let _ = <ActiveContainerChainsForCurrentSession<T>>::try_mutate(
-                |active_chains| -> DispatchResult {
-                    if active_chains.try_insert(chain_id).is_err() {
-                        // If we reach MaxContainerChains limit there must be a bug in the pallet
-                        // so we disable the activity tracking
-                        Self::set_inactivity_tracking_status_inner(
-                            T::CurrentSessionIndex::session_index(),
-                            false,
-                        );
-                        total_weight.saturating_accrue(T::DbWeight::get().reads_writes(1, 2));
-                        return Err(Error::<T>::MaxContainerChainsReached.into());
-                    } else {
-                        total_weight += T::DbWeight::get().writes(1);
+        pub fn on_chains_noted(chains: impl IntoIterator<Item = ParaId>) -> Weight {
+            let mut total_weight = T::DbWeight::get().reads(1);
+
+            let result = <ActiveContainerChainsForCurrentSession<T>>::try_mutate(|active_chains| {
+                let mut temp_set: BTreeSet<ParaId> = active_chains.clone().into();
+
+                for chain in chains {
+                    temp_set.insert(chain);
+                }
```

Comment:
Maybe there is some api to merge 2 btreesets?

Reply:
@tmpolaczyk , check how we don't even need temp vars, WDYT?
https://github.com/moondance-labs/tanssi/pull/1179/commits/7a1fe8bb0a57a693a5507fab14816483961fae86
It looks cleaner and more concise.

Reply:
It's shorter but the previous code was much easier to understand, without comments I won't know what is this in a few months.

If you want to avoid the tmp variable, maybe try this method? https://paritytech.github.io/polkadot-sdk/master/sp_core/bounded_btree_map/struct.BoundedBTreeMap.html#method.try_mutate

Reply:
Reverted here: https://github.com/moondance-labs/tanssi/commit/05f7de88b700f98422dbffb550b83ee996e4dc73